src/gt4py/gtc/python/oir_to_npir.py

```diff
@@ -0,0 +1,104 @@
+from typing import Dict, List, Optional
+
+from dataclasses import dataclass, field
+from eve.visitors import NodeTranslator
+
+from .. import common, oir
+from . import npir
+
+
+class OirToNpir(NodeTranslator):
+    """Lower from optimizable IR (OIR) to numpy IR (NPIR)."""
+
+    @dataclass
+    class Context:
+        """Context for a HorizontalExecution."""
+
+        domain_padding: Dict[str, List] = field(
+            default_factory=lambda: {"lower": [0, 0, 0], "upper": [0, 0, 0]}
+        )
+        parallel_k: Optional[bool] = None
+
+        def set_parallel_k(self, value: bool) -> "OirToNpir.Context":
+            return self.__class__(parallel_k=value, domain_padding=self.domain_padding)
+
+        def add_offset(self, axis: int, value: int) -> "OirToNpir.Context":
+            if value > 0:
+                current_value = self.domain_padding["upper"][axis]
+                self.domain_padding["upper"][axis] = max(current_value, value)
+            elif value < 0:
+                current_value = self.domain_padding["lower"][axis]
+                self.domain_padding["lower"][axis] = max(current_value, abs(value))
+            return self
+
+        def add_offsets(self, i: int, j: int, k: int) -> "OirToNpir.Context":
+            for axis_index, value in enumerate([i, j, k]):
+                self.add_offset(axis_index, value)
+            return self
+
+    def visit_Stencil(
+        self, node: oir.Stencil, *, ctx: Optional[Context] = None
```

Comment:
I think ctx shouldn't be a parameter here. If it is then it `domain_padding` needs to be reset when entering here (or before leaving) as the accumulated padding is not valid outside of the scope of this stencil. Currently it doesn't matter because it's the root node, but currently it's also not needed to have it as a parameter. I'd rather remove it.

If there is more context than just padding, we maybe should have different ctx objects for the different scopes.

Reply:
The advantage of the parameter is that assertions can be run on the context afterward in tests. Otherwise you need to return the context from the visitor...